spark/spark-tensorflow-distributor/spark_tensorflow_distributor/mirrored_strategy_runner.py

```diff
@@ -0,0 +1,331 @@
+                f'spark.task.resource.{gpu_resource_name}, '
+                f'spark.worker.resource.{gpu_resource_name}, '
+                f'spark.executor.resource.{gpu_resource_name}, and '
+                f'spark.driver.resource.{gpu_resource_name} confs are set and that the '
+                f'GPU resource name `{gpu_resource_name}` matches those confs correctly.'
+            )
+
+    # Runs the training function
+    @staticmethod
+    def _run_tensorflow_program(train_fn, use_custom_strategy, **kwargs):
+        if use_custom_strategy:
+            return train_fn(**kwargs)
+        else:
+            import tensorflow as tf
+            from tensorflow.python.eager import context
+            context._reset_context()
+            strategy = tf.distribute.experimental.MultiWorkerMirroredStrategy()
+            with strategy.scope():
+                return train_fn(**kwargs)
+
+    def _get_spark_task_program(self, train_fn, **kwargs):
+        num_slots = self.num_slots
+        use_custom_strategy = self.use_custom_strategy
+        gpu_resource_name = self.gpu_resource_name
+        num_tasks = self.num_tasks
+        is_gpu = self.is_gpu
+        get_gpus_owned = MirroredStrategyRunner._get_gpus_owned
+        run_tensorflow_program = MirroredStrategyRunner._run_tensorflow_program
+
+        # Spark task program
+        def wrapped_train_fn(_):
+            import json, logging, os, socket
+            from contextlib import closing
+            from pyspark import BarrierTaskContext
+
+            # Sets the TF_CONFIG env var so TF servers can communicate with each other
+            def set_tf_config(context):
+                addrs = [e.address.split(':')[0] for e in context.getTaskInfos()]
+                my_addr = addrs[context.partitionId()]
+                with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as my_sock:
+                    my_sock.bind(('', 0))
+                    _, my_port = my_sock.getsockname()
+                    my_endpoint = "{}:{}".format(my_addr, my_port)
+                    worker_endpoints = context.allGather(my_endpoint)
+                cluster = {
+                    'worker': worker_endpoints
+                }
+                tf_config = {
+                    'cluster': cluster,
+                    'task': {
+                        'type': 'worker',
+                        'index': context.partitionId()
+                    }
+                }
+                os.environ['TF_CONFIG'] = json.dumps(tf_config)
+
+            # Sets the CUDA_VISIBLE_DEVICES env var so only the appropriate GPUS are used
+            def set_gpus(context):
+                gpus_owned = get_gpus_owned(context.resources(), gpu_resource_name)
+                my_num_gpus = (num_slots // num_tasks) + (context.partitionId() < (num_slots % num_tasks))
+                gpu_addresses = [str(e) for e in random.sample(gpus_owned, my_num_gpus)]
```

Comment:
Can I assume that all GPUs on a machine will be taken except for the last machine? The reason I ask is randomly chosen GPUs may not have NVLinks in between.

Reply:
Ah good point, we can't. For example, if we do GPU training with 6 GPUs (`num_slots=6`) on a 2 worker cluster, each with 4 GPUs, then the training will be done with 3 GPUs on each worker, rather than 4 GPUs on one worker and 2 GPUs on the other. Currently Spark's resource scheduling doesn't support resource groups, so I'd suggest leaving this as a TODO until we can use that Spark feature to report the NVLink groups to Spark.